File: benchmarks/baselines/go/v1/jobservice.go
// Package v1 is the version-one checkpoint for the UC-001 implementation task.
package v1

import (
	"unicode"
	"unicode/utf8"
)

// CreateJobRequest is the decoded version-one public request.
type CreateJobRequest struct {
	JobID   string
	Task    string
	Payload []byte
}

// Job is the version-one record accepted by the jobs store.
type Job struct {
	JobID   string
	Task    string
	Payload []byte
}

type ValidationField string

const (
	FieldJobID   ValidationField = "job_id"
	FieldTask    ValidationField = "task"
	FieldPayload ValidationField = "payload"
)

type ValidationReason string

const (
	ReasonMissing          ValidationReason = "missing"
	ReasonInvalidFormat    ValidationReason = "invalid_format"
	ReasonTooLong          ValidationReason = "too_long"
	ReasonControlCharacter ValidationReason = "control_character"
	ReasonPayloadTooLarge  ValidationReason = "payload_too_large"
)

type ValidationIssue struct {
	Field  ValidationField
	Reason ValidationReason
}

// CreateJobResult is the closed public result contract for UC-001.
type CreateJobResult interface {
	isCreateJobResult()
}

type Created struct{ Job Job }
type Invalid struct{ Issues []ValidationIssue }
type AlreadyExists struct{ JobID string }
type PersistenceUnavailable struct{}

func (Created) isCreateJobResult()                {}
func (Invalid) isCreateJobResult()                {}
func (AlreadyExists) isCreateJobResult()          {}
func (PersistenceUnavailable) isCreateJobResult() {}

type InsertOutcome uint8

const (
	Inserted InsertOutcome = iota
	Duplicate
	UnavailableBeforeCommit
)

// JobStore is the only capability available to CreateJob.
type JobStore interface {
	InsertIfAbsent(Job) InsertOutcome
}

// CreateJob validates the complete request before making exactly one store call.
func CreateJob(request CreateJobRequest, store JobStore) CreateJobResult {
	issues := validate(request)
	if len(issues) != 0 {
		return Invalid{Issues: issues}
	}

	job := Job{
		JobID:   request.JobID,
		Task:    request.Task,
		Payload: request.Payload,
	}
	switch store.InsertIfAbsent(job) {
	case Inserted:
		return Created{Job: job}
	case Duplicate:
		return AlreadyExists{JobID: job.JobID}
	case UnavailableBeforeCommit:
		return PersistenceUnavailable{}
	default:
		panic("job store returned an invalid insert outcome")
	}
}

func validate(request CreateJobRequest) []ValidationIssue {
	issues := make([]ValidationIssue, 0, 3)

	switch {
	case request.JobID == "":
		issues = append(issues, ValidationIssue{Field: FieldJobID, Reason: ReasonMissing})
	case !validJobID(request.JobID):
		issues = append(issues, ValidationIssue{Field: FieldJobID, Reason: ReasonInvalidFormat})
	}

	switch {
	case request.Task == "":
		issues = append(issues, ValidationIssue{Field: FieldTask, Reason: ReasonMissing})
	case utf8.RuneCountInString(request.Task) > 80:
		issues = append(issues, ValidationIssue{Field: FieldTask, Reason: ReasonTooLong})
	case containsControl(request.Task):
		issues = append(issues, ValidationIssue{Field: FieldTask, Reason: ReasonControlCharacter})
	}

	if len(request.Payload) > 4096 {
		issues = append(issues, ValidationIssue{Field: FieldPayload, Reason: ReasonPayloadTooLarge})
	}
	return issues
}

func validJobID(jobID string) bool {
	if len(jobID) == 0 || len(jobID) > 64 || !isASCIIAlphanumeric(jobID[0]) {
		return false
	}
	for index := 1; index < len(jobID); index++ {
		value := jobID[index]
		if !isASCIIAlphanumeric(value) && value != '_' && value != '-' {
			return false
		}
	}
	return true
}

func isASCIIAlphanumeric(value byte) bool {
	return value >= 'a' && value <= 'z' ||
		value >= 'A' && value <= 'Z' ||
		value >= '0' && value <= '9'
}

func containsControl(value string) bool {
	for _, current := range value {
		if unicode.IsControl(current) {
			return true
		}
	}
	return false
}

File: benchmarks/baselines/typescript/v1/job-service.ts
/** Version-one UC-001 checkpoint. */

export type CreateJobRequest = Readonly<{
  jobId: string;
  task: string;
  payload: Uint8Array;
}>;

export type Job = Readonly<{
  jobId: string;
  task: string;
  payload: Uint8Array;
}>;

export type ValidationField = "job_id" | "task" | "payload";
export type ValidationReason =
  | "missing"
  | "invalid_format"
  | "too_long"
  | "control_character"
  | "payload_too_large";

export type ValidationIssue = Readonly<{
  field: ValidationField;
  reason: ValidationReason;
}>;

export type CreateJobResult =
  | Readonly<{ kind: "created"; job: Job }>
  | Readonly<{ kind: "invalid"; issues: readonly ValidationIssue[] }>
  | Readonly<{ kind: "already_exists"; jobId: string }>
  | Readonly<{ kind: "persistence_unavailable" }>;

export const InsertOutcome = {
  inserted: "inserted",
  duplicate: "duplicate",
  unavailableBeforeCommit: "unavailable_before_commit",
} as const;

export type InsertOutcome = (typeof InsertOutcome)[keyof typeof InsertOutcome];

export type JobStore = {
  /** The only external capability available to createJob. */
  insertIfAbsent(job: Job): InsertOutcome;
};

const jobIdPattern = /^[A-Za-z0-9][A-Za-z0-9_-]{0,63}$/u;
const controlCharacterPattern = /\p{Cc}/u;

export function createJob(
  request: CreateJobRequest,
  store: JobStore,
): CreateJobResult {
  const issues = validate(request);
  if (issues.length > 0) {
    return { kind: "invalid", issues };
  }

  const job: Job = {
    jobId: request.jobId,
    task: request.task,
    payload: Uint8Array.from(request.payload),
  };
  const outcome = store.insertIfAbsent(job);
  switch (outcome) {
    case InsertOutcome.inserted:
      return { kind: "created", job };
    case InsertOutcome.duplicate:
      return { kind: "already_exists", jobId: job.jobId };
    case InsertOutcome.unavailableBeforeCommit:
      return { kind: "persistence_unavailable" };
    default:
      return assertNever(outcome);
  }
}

function validate(request: CreateJobRequest): readonly ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  if (request.jobId.length === 0) {
    issues.push({ field: "job_id", reason: "missing" });
  } else if (!jobIdPattern.test(request.jobId)) {
    issues.push({ field: "job_id", reason: "invalid_format" });
  }

  if (request.task.length === 0) {
    issues.push({ field: "task", reason: "missing" });
    // Array.from counts Unicode code points, which are the fixture's scalar unit.
  } else if (Array.from(request.task).length > 80) {
    issues.push({ field: "task", reason: "too_long" });
  } else if (controlCharacterPattern.test(request.task)) {
    issues.push({ field: "task", reason: "control_character" });
  }

  if (request.payload.byteLength > 4096) {
    issues.push({ field: "payload", reason: "payload_too_large" });
  }
  return issues;
}

function assertNever(value: never): never {
  throw new Error(`invalid insert outcome: ${String(value)}`);
}
